R/tidy_output.R

```diff
@@ -0,0 +1,318 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# .make_token_map ---------------------------------------------------------
+
+
+#' Make Token Map
+#'
+#' Given a list of input token sequences, construct a data frame with explicit
+#' index columns. Later this can be joined onto the indexed outputs to add the
+#' actual token and segment columns to the tidy data frames.
+#'
+#' @inheritParams tidy_bert_output
+#'
+#' @return A data frame of the input tokens, with explicit index columns.
+#' @keywords internal
+.make_token_map <- function(tokenized) {
+  tok_id_list <- tokenized$token_ids
+  tok_list <- purrr::map(tok_id_list, names)
+  tt_list <- tokenized$token_type_ids
+
+  token_map <- tibble::tibble(
+    sequence_index = integer(),
+    token_index = integer(),
+    segment_index = integer(),
+    token = character()
+  )
+  for (i in seq_along(tok_list)) {
+    tokens <- tok_list[[i]]
+    token_ids <- tok_id_list[[i]]
+    segments <- tt_list[[i]]
+    token_seq <- seq_along(tokens)
+    token_map <- dplyr::bind_rows(
+      token_map,
+      data.frame(
+        sequence_index = as.integer(i),
+        token_index = token_seq,
+        segment_index = segments,
+        token = tokens,
+        token_id = token_ids,
+        stringsAsFactors = FALSE
+      )
+    )
+  }
+
+  return(token_map)
+}
+
+
+
+# .make_index_column ------------------------------------------------------
+
+
+#' Make Index Column
+#'
+#' Given an overall shape and a particular dimension, construct the index column
+#' for that dimension.
+#'
+#' @param shape Integer vector; the shape of the tensor to index.
+#' @param which Integer scalar; which dimension to index.
+#'
+#' @return The index column as an integer vector.
+#' @keywords internal
+.make_index_column <- function(shape, which) {
+  temp <- torch::torch_ones(dim = shape)
+
+  dim <- rep(1, length(shape))
+  dim[[which]] <- shape[[which]]
+  index_tensor <- torch::torch_tensor(array(
+    seq_len(shape[[which]]),
+    dim = dim
+  ))
+
+  return(as.integer(torch::torch_flatten(
+    temp * index_tensor, # broadcast
+    start_dim = 1,
+    end_dim = length(shape)
+  )))
+}
+# .process_attention_result -----------------------------------------------
+
+
+#' Process Attention
+#'
+#' Takes the raw output from a BERT model and turns the attention weights matrix
+#' into a data frame.
+#'
+#' @inheritParams tidy_bert_output
+#'
+#' @return The attention weights from a BERT model, as a data frame.
+#' @keywords internal
+.process_attention_result <- function(bert_model_output) {
+  attention_output <- bert_model_output$attention_weights
+
```

Comment:
Should we validate that the output has the shape we expect? Or do that higher up (after adding some classing to {tt} output and then trust it?

Reply:
If we're going to add classing to {tt} output at some point, then I think we let that handle the validation for us. In that case, we could implement this as a method. 🤔  For this iteration, I'm more concerned with getting the capability up. Better error messaging can come in future iterations.